Add a blog post for the release of opam 2.4.0~alpha1 (#111)
ocaml/platform-blog: opam-2-4-0-alpha1.md

]
date: "2025-04-17"
--BODY--

We are happy to announce the first alpha release of opam 2.4.0.
You can view the full list of changes in the
[release note](https://github.com/ocaml/opam/releases/tag/2.4.0-alpha1).

This version is an alpha, we invite users to test it to spot previously
unnoticed bugs as we head towards the stable release.

## Try it!

The upgrade instructions are unchanged:

1. Either from binaries: run

For Unix systems
```
bash -c "sh <(curl -fsSL https://opam.ocaml.org/install.sh) --version 2.4.0~alpha1"
```
or from PowerShell for Windows systems
```
Invoke-Expression "& { $(Invoke-RestMethod https://opam.ocaml.org/install.ps1) } -Version 2.4.0~alpha1"
```
or download manually from [the Github "Releases" page](https://github.com/ocaml/opam/releases/tag/2.4.0-alpha1) to your PATH.

2. Or from source, manually: see the instructions in the [README](https://github.com/ocaml/opam/tree/2.4.0-alpha1#compiling-this-repo).

You should then run:
```
opam init --reinit -ni
```

## Major changes

* Remove `ocaml-system` from the list of default compilers chosen at `opam init` time ([#3509](https://github.com/ocaml/opam/issues/3509))

> * Patches are now applied using the `patch` OCaml library instead of GNU Patch ([#6019](https://github.com/ocaml/opam/issues/6019) [#6052](https://github.com/ocaml/opam/issues/6052) [#3782](https://github.com/ocaml/opam/issues/3782) [ocaml/setup-ocaml#933](https://github.com/ocaml/setup-ocaml/pull/933))

Author: ```suggestion
* Patches are now applied using the `patch` OCaml library instead of GNU Patch ([#6019](https://github.com/ocaml/opam/issues/6019), [#6052](https://github.com/ocaml/opam/issues/6052), [#3782](https://github.com/ocaml/opam/issues/3782), [ocaml/setup-ocaml#933](https://github.com/ocaml/setup-ocaml/pull/933))
```

  * Context diffs are not supported anymore, only Unified diffs are (including its git extensions)
  * Stop support of file permission changes via git extension to the unified diff specification
  * GNU `patch` and the `diff` command are no longer runtime dependencies

* Support providing external dependencies with Nix by adding support for stateless depexts systems ([#5982](https://github.com/ocaml/opam/issues/5982)). *Thanks to [@RyanGibb](https://github.com/RyanGibb) for this contribution*

> * Fix `opam install <local_dir>` not updating or storing pinned packages' metadata ([6248]())

Author: ```suggestion
* Fix `opam install <local_dir>` not updating or storing pinned packages' metadata ([#6248](https://github.com/ocaml/opam/issues/6248))
```

> * Fix `opam install <local_dir>` not updating or storing pinned packages' metadata ([#6248](https://github.com/ocaml/opam/issues/6248))

Reviewer: ```suggestion
* Fix `opam install <local_dir>` not updating or storing pinned packages' metadata ([#6248](https://github.com/ocaml/opam/issues/6248), [#6379](https://github.com/ocaml/opam/issues/6379))
```

> * Fix `opam install --deps-only/--show-action <local_dir>` not updating (without storing) pinned packages' metadata ([#5567](https://github.com/ocaml/opam/issues/5567))

Author: i'm not really satisfied with these. They don't really explain what we've fixed from a user PoV, only from our biased view over the opam internals

Author: i don't really know what to say instead though. Do you have an idea?

Reviewer: Following what users were seeing (and reftests), maybe something like

* Fix `opam install <local_dir>` (and with options) not taking into account dependencies change in the opam file ([#6248](https://github.com/ocaml/opam/issues/6248))

* Fix `opam install --deps-only/--show-action <local_dir>` reporting *Nothing to do* ([#5567](https://github.com/ocaml/opam/issues/5567))

Author: I like the line for `--deps-only`, but the first one the issue is different. It's the fact that opam doesn't store the metadata of the package its installing and instead might be relying on the definition from opam-repository, which mess things up for every other commands basically

> * Fix `opam install --deps-only/--show-action <local_dir>` not updating (without storing) pinned packages' metadata ([#5567](https://github.com/ocaml/opam/issues/5567))

Reviewer: ```suggestion
* Fix `opam install --deps-only/--show-action <local_dir>` not updating (without storing) pinned packages' metadata ([#5567](https://github.com/ocaml/opam/issues/5567), [#4507](https://github.com/ocaml/opam/issues/4507))
```

> * Remove `ocaml-system` from the list of default compilers chosen at `opam init` time ([#3509](https://github.com/ocaml/opam/issues/3509))
> 
> * Patches are now applied using the `patch` OCaml library instead of GNU Patch ([#6019](https://github.com/ocaml/opam/issues/6019), [#6052](https://github.com/ocaml/opam/issues/6052), [#3782](https://github.com/ocaml/opam/issues/3782), [ocaml/setup-ocaml#933](https://github.com/ocaml/setup-ocaml/pull/933))
>   * Context diffs are not supported anymore, only Unified diffs are (including its git extensions)
>   * Stop support of file permission changes via git extension to the unified diff specification
>   * GNU `patch` and the `diff` command are no longer runtime dependencies
> 
> * Support providing external dependencies with Nix by adding support for stateless depexts systems ([#5982](https://github.com/ocaml/opam/issues/5982)). *Thanks to [@RyanGibb](https://github.com/RyanGibb) for this contribution*
> 
> * Fix `opam install <local_dir>` not updating or storing pinned packages' metadata ([#6248](https://github.com/ocaml/opam/issues/6248), [#6379](https://github.com/ocaml/opam/issues/6379))
> 
> * Fix `opam install --deps-only/--show-action <local_dir>` not updating (without storing) pinned packages' metadata ([#5567](https://github.com/ocaml/opam/issues/5567), [#4507](https://github.com/ocaml/opam/issues/4507))

Author: ```suggestion
* On `opam init` the compiler chosen for the default switch will no longer be `ocaml-system` ([#3509](https://github.com/ocaml/opam/issues/3509))
  This was done because the system compiler (as-is your ocaml installed system wide, e.g. /usr/bin/ocaml) is known to be under-tested and prone to a variety of bugs and configuration issues.
  Removing it from the default compiler allows new-comers a more smooth experience.
  *Note: if you wish to use it anyway, you are always able to do it explicitly using `opam init --compiler=ocaml-system`*

* GNU `patch` and the `diff` command are no longer runtime dependencies. Instead the OCaml `patch` library is used ([#6019](https://github.com/ocaml/opam/issues/6019), [#6052](https://github.com/ocaml/opam/issues/6052), [#3782](https://github.com/ocaml/opam/issues/3782), [ocaml/setup-ocaml#933](https://github.com/ocaml/setup-ocaml/pull/933))
  Doing this we've removed some rarely used features of GNU Patch such as the support of [Context diffs](https://www.gnu.org/software/diffutils/manual/html_node/Example-Context.html).
  The new implementation only supports [Unified diffs](https://www.gnu.org/software/diffutils/manual/html_node/Example-Unified.html) including the [git extended headers](https://git-scm.com/docs/diff-format), however file permission changes via said extended headers have no effect.

* Add Nix support for external dependencies (depexts) by adding support for stateless package managers ([#5982](https://github.com/ocaml/opam/issues/5982)). *Thanks to [@RyanGibb](https://github.com/RyanGibb) for this contribution*

* **Fix `opam install <local_dir>`** with and without options like `--deps-only` or `--show-action` having **unexpected behaviours** (([#6248](https://github.com/ocaml/opam/issues/6248)), [#5567](https://github.com/ocaml/opam/issues/5567)) such as:
  * reporting `Nothing to do` despite dependencies or package not being up-to-date
  * asking to install the wrong dependencies
```

> 

Author: ```suggestion

## Other noteworthy changes

```

* `opam upgrade`: Do not show the message about packages "not up-to-date" when the package is tagged with `avoid-version`/`deprecated` ([#6271](https://github.com/ocaml/opam/issues/6271))

* Fail when trying to pin a package whose definition could not be found instead of forcing interactive edition (e.g. this could happen when making a typo in the package name of a pin-depends) ([#6322](https://github.com/ocaml/opam/issues/6322))

## New commands / options

* Add `opam admin compare-versions` to compare package versions for sanity checks. *Thanks to [@mbarbin](https://github.com/mbarbin) for this contribution*

* Add `opam lock --keep-local` to keep local pins url in `pin-depends` field ([#4897](https://github.com/ocaml/opam/issues/4897))

* Add `opam admin migrate-extrafiles` which moves all `extra-files` of an existing opam repository into `extra-sources`. *Thanks to [@hannesm](https://github.com/hannesm) for this contribution*

* The `-i`/`--ignore-test-doc` argument has been removed from `opam admin check` ([#6335](https://github.com/ocaml/opam/issues/6335))

> Various performance and other improvements were made and bugs were fixed.
> API changes are denoted in the release note linked above.

Reviewer: ```suggestion
Various performance and other improvements were made and bugs were fixed.
API changes are also denoted in the release note linked above.
```
I don't know if it is the best to put this "also" here. The idea is to say to user that the "performance & improvments" and "api changes" are in the release, but without putting "in the release note" in the two sentences.
Maybe
```suggestion
Various performance and other improvements were made and bugs were fixed, you can find the full list in the release note linked above. API changes are denoted there.
```
but it also doubles the information with the beginning of the thread.

Author: The "also" is fine there, i've added it. I'm not sure about the second suggestion though. I don't really like the idea of making it one big line and also having to read the previous sentence if you're just interested the API.
